[XeGPU] uArch definition (PR 1/N)

mshahneo opened the pull request:
Update 2:
---------
After another round of review update the uArch definition in the following way:
- Add common 5 APIs for all the interfaces.
- Make OpInterfaces more specific (e.g., 2D, 1D Block IO have seperate interfaces).
- Make the design more specific to Intel hardware.
- Remove the information we don't use, i.e., make the design simpler.

Update:
---------
After the first iteration of review comment, the current PR is update in the following way:

- The main pivot for this iteration is the utilities. These utilities are expose via a range of Interfaces (e.g. TileOpInterface, MatrixOpInterface). A specific instruction for a specific Architecture need to implement theses utilities.
- The uArch information is now directly embedded in the utility implementation. We are not using .yaml files. However, I haven't removed the .yaml file yet, since, they can still be utilized if choses to do so in the future.
- Also, last but not least, the current version provides an end-to-end prototype of how uArch can be used:
--  Full uArch definition, utilities and all (at least for one instruction for now, DPAS).
--  Definition for 2 uArchs (PVC, BMG).
-- A pass to attach the uArch name to the module attribute using DLTI.
-- Shows the usage of the uArch interface to verify XeGPU ops.
-- Few test cases.

**P.S.** I still kept some old code intentionally, if we want to go back or use some of it. I'll clean up once we agree on the process.

================================

Hi All,

This is the initial PR for uArch definition. The primary purpose for this PR is to have a discussion and finalize the design. So please think of it as an RFC as well.

First target we want to achieve here is to finalize the skeleton:

- Skeleton of the YAML file, what information we want in there
- Skeleton of Data structures we want to have (e.g., necessary to create utilities to support the need)

Due to this purpose, the main focus point of this PR are the following files:

- `mlir/lib/Dialect/XeGPU/Utils/intel_gpu_pvc.yaml `: holds the YAML definition of uArch (PVC is used for this use case)
- `mlir/include/mlir/Dialect/XeGPU/Utils/uArch.h` : holds the base uArch struct that can be re-used by specialized uArchs such as PVC.
- `mlir/include/mlir/Dialect/XeGPU/Utils/IntelGpuPVC.h` : holds the PVC-specific structs.

mlir/lib/Dialect/XeGPU/Utils/IntelGpuPVC.cpp is added as placeholder for YAML mapping. This file is intentionally not complete. We want to reach a consensus on the YAML, and C++ structures before filling this one in.

Please provide your valuable feedback as we try to finalize this together.

mshahneo committed:
[uArch][XeGPU] Add uArch definition.

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Utils/IntelGpuPVC.h:
```diff
@@ -0,0 +1,100 @@
+} // namespace mlir
+
+#endif // MLIR_DIALECT_XEGPU_UTILS_INTEL_GPU_PVC_H
+//===--- IntelGpuPVC.h ---------------------------------------*- C++ -*-===//
```
this one should be on the top I guess

mshahneo replied:
Ah, yes, I think I auto-added by co-pilot suggestion. Will remove it :)

adam-smnk commented:
A general question on the structure - what's the benefit of maintaining both yaml and C++ definitions?
Do we expect sth to reuse yaml separately from C++? Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?

adam-smnk commented on mlir/include/mlir/Dialect/XeGPU/Utils/IntelGpuPVC.h:
```diff
@@ -0,0 +1,100 @@
+  Range systolic_depth;
+  Range repreat_count;
+  Range execution_size;
+  std::map<std::string, uint> ops_per_channel;
```
Opaque string mapping doesn't feel user friendly. Especially as it seems to be wrapping numerical bit width that could be a numerical key directly - I assume it's some limitation due to using (conversion from) yaml.

`matrix_size` is even worse offender in that regard. These wrappers are hard to use without yaml definitions which again makes me wonder if we really need to split it into two separate files.

mshahneo replied:
> Opaque string mapping doesn't feel user friendly. Especially as it seems to be wrapping numerical bit width that could be a numerical key directly - I assume it's some limitation due to using (conversion from) yaml.

You are right, the limitation is due to the yaml mapping.

> matrix_size is even worse offender in that regard. These wrappers are hard to use without yaml definitions which again makes me wonder if we really need to split it into two separate files.

Sorry, the matrix size is a mistake. The value should be a vector of uint. But yes, in general you are right, the structs and and yaml to some extent have to be used in conjunction.

mshahneo commented:
> A general question on the structure - what's the benefit of maintaining both yaml and C++ definitions? Do we expect sth to reuse yaml separately from C++? Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?

The C++ structure actually comes out of a necessity. To use yaml mapping utlity in LLVM, we have to have a C++ object mapped to it, hence the structures. And since we need the C++ structures anyway, I wanted to make some of them to be re-usable across uArchs.

> Do we expect sth to reuse yaml separately from C++

Not really, not in this context anyway. But we wanted to keep the base structs in a such a way that they can be used in a non-LLVM cases, but that's more of hope not necessasity.

> Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?

I thought about this, but it takes me back to tablegen. Should we use tablegen?

adam-smnk commented:
> > A general question on the structure - what's the benefit of maintaining both yaml and C++ definitions? Do we expect sth to reuse yaml separately from C++? Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?
> 
> The C++ structure actually comes out of a necessity. To use yaml mapping utlity in LLVM, we have to have a C++ object mapped to it, hence the structures. And since we need the C++ structures anyway, I wanted to make some of them to be re-usable across uArchs.
> 
> > Do we expect sth to reuse yaml separately from C++
> 
> Not really, not in this context anyway. But we wanted to keep the base structs in a such a way that they can be used in a non-LLVM cases, but that's more of hope not necessasity.

It feels like the overhead of this approach in its complexity and possible maintenance cost is not really justified. On its own it seems fine but these C++ bindings are not great. 😅
Overall, I'd say yaml is not really a first-class citizen in MLIR so it'd face extra scrutiny during upstreaming.

> > Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?
> 
> I thought about this, but it takes me back to tablegen. Should we use tablegen?

It's worth a try if what we need easily translates to tablegen entries and structure. That is, if you find yourself at any point fighting against tablegen or in need to introduce custom "hacks", I would not necessarily double down on it.

I would suggest starting with user interfaces to flesh out what's really needed. Then it should become clearer where to store all the uarch details e.g., embed shared memory size directly in `getShmemSize()` or create a separate struct or pull it from a constexpr dictionary that contains everything etc.

Next quick test would be to see how well it all holds up for another target like battlemage.

mshahneo commented:
> > > A general question on the structure - what's the benefit of maintaining both yaml and C++ definitions? Do we expect sth to reuse yaml separately from C++? Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?
> > 
> > 
> > The C++ structure actually comes out of a necessity. To use yaml mapping utlity in LLVM, we have to have a C++ object mapped to it, hence the structures. And since we need the C++ structures anyway, I wanted to make some of them to be re-usable across uArchs.
> > > Do we expect sth to reuse yaml separately from C++
> > 
> > 
> > Not really, not in this context anyway. But we wanted to keep the base structs in a such a way that they can be used in a non-LLVM cases, but that's more of hope not necessasity.
> 
> It feels like the overhead of this approach in its complexity and possible maintenance cost is not really justified. On its own it seems fine but these C++ bindings are not great. 😅 Overall, I'd say yaml is not really a first-class citizen in MLIR so it'd face extra scrutiny during upstreaming.
> 
> > > Can the C++ bindings be autogenerated or will it be a lot of manual boilerplate?
> > 
> > 
> > I thought about this, but it takes me back to tablegen. Should we use tablegen?
> 
> It's worth a try if what we need easily translates to tablegen entries and structure. That is, if you find yourself at any point fighting against tablegen or in need to introduce custom "hacks", I would not necessarily double down on it.
> 
> I would suggest starting with user interfaces to flesh out what's really needed. Then it should become clearer where to store all the uarch details e.g., embed shared memory size directly in `getShmemSize()` or create a separate struct or pull it from a constexpr dictionary that contains everything etc.
> 
> Next quick test would be to see how well it all holds up for another target like battlemage.

Thanks, Adam. I think that's a good idea. Let me try to collect the uArch info battlemage. I looked into them during the design, but more details may help us finding the approach that would be easy to maintain.

mshahneo committed:
Address review comments.

mshahneo committed:
Modify the uArch definition.

This version focuses on the utilities to be the pivot.
It also saves info directly in C++ files as part of get functions.
Don't use the yamls anymore.

Adds support for DPAS instruction.

mshahneo commented:
Hi Adam (@adam-smnk), Igor (@Garra1980 ),

Modified the uarch definition based on our discussion.
In this version the Pivot is the utilities which necessary instructions have to implement.
I also moved the information to C++ (as part of get functions directly instead of in yaml).

Please let me know what you think. I only added implementation for DPAS. Will add Load/store/prefetch if we agree on the process.

Thanks a lot in advance :).

P.S. I still kept some old code intentionally, if we want to go back or use some of it. I'll clean up once we agree on the process.

mshahneo commented:
@silee2 , @nbpatel , @chencha3 , @charithaintc , @Jianhui-Li, @akroviakov 
please let me know if you have any comments/suggestions.
Thank you so much in advance :)

Garra1980 commented:
also cc @tkarna

tkarna commented:
Do you think xegpu could at some point support the Data Layout and Target Information (DLTI) dialect for querying the uarch info?
FYI @rolfmorel

rolfmorel commented:
Thanks for adding me, @tkarna ! Agreed, it would be great to collaborate on ways to expose this target info through DLTI. 

I think @tkarna's work on performant schedules for Xe - which have plentiful magic values derivable from hw info - should be a good exemplar of a user of Xe's DLTI. Showing how to automatically derive those values through DLTI would be a great success indicator.

I had a quick attempt at what exposing this data through DLTI might look like (here using the transform dialect though you can also do the same queries from C++):

```mlir
module attribute { #xevm.target<arch = "intel_gpu_pvc", ... = ?any other flags that vary across hardware?> } {
    ...
}
module {transform.with_named_sequence} {
    transform.named_sequence @__main__(%mod: !transform.any_op) {
        %number_of_eus_per_xe_core = transform.dlti.query ["xe_core","#eus"] at %mod  // #xevm.target answers 8
        %number_of_threads_per_eu = transform.dlti.query ["xe_core","eu","#threads"] at %mod  // #xevm.target answers [8, 4]
        
        %shared_memory_size = transform.dlti.query ["xe","shared_memory","size"] at %mod  // #xevm.target answers 524288
        %shared_memory_alignment = transform.dlti.query ["xe","shared_memory","alighment"] at %mod  // #xevm.target answers 64

        %grf_bitwidth = transform.dlti.query ["xe","grf","bitwidth"] at %mod  // #xevm.target answers 512
        %grf_modes = transform.dlti.query ["xe","grf","modes"] at %mod  // #xevm.target answers ["small", "large"]
        %grf_count_per_thread = transform.dlti.query ["xe","grf","mode","small","per_thread"] at %mod  // #xevm.target answers 128

        %dpas_opcode = transform.dlti.query ["xe","inst","dpas","opcode"] at %mod  // #xevm.target answers 0x83
        %dpas_systolic_depth = transform.dlti.query ["xe","inst","dpas","systolic_depth"] at %mod  // #xevm.target answers [8]
        %dpas_repeat_count = transform.dlti.query ["xe","inst","dpas","repeat_count"] at %mod  // #xevm.target answers 1,2,3,4,5,6,7,8 as multiple associations for single result handle OR [1, 8] OR #dlti.map<"min" = 1, "max" = 8> OR #dlti.range<1,2,...,8>/#tune.range<1,2,...,8> which would also respond to "min" and "max" queries,
        %dpas_ops_per_channel = transform.dlti.query ["xe","inst","dpas","ops_per_channel"] at %mod  // #xevm.target answers #dlti.map<19: 1, 16: 2, 8: 4, 4: 8, 2:8>
        
        %dpas_supported_types = transform.dlti.query ["xe","inst","dpas","supported_types"] at %mod  // #xevm.target answers #xevm.dpas_types so that ...
        %dpas_supported_types_B_wildcard = transform.dlti.query ["xe","inst","dpas","supported_types", f16, f16, f16, #dlti._] at %mod  // #xevm.dpas_types (yielded by #xevm.target which answered prefix) answers f16
        %dpas_supported_types_Dst_wildcard = transform.dlti.query ["xe","inst","dpas","supported_types", #dlti._, f16, f16, f16] at %mod  // #xevm.dpas_types (yielded by #xevm.target which answered prefix) answers f16,f32 (i.e. two type params associated to the handle)
        %dpas_supported_types_Dst_and_Acc_wildcard:2 = transform.dlti.query ["xe","inst","dpas","supported_types", #dlti._, #dlti._, f16, f16] at %mod  // #xevm.dpas_types (yielded by #xevm.target which answered prefix) answers f16,f32,f16,f32 for first result and f16,f16,f32,f32 for second result
        
        ... // transform ops would take the above Values as parameters
    }
}
```

That is, there's an `#xevm.target` attribute (or `#xegpu.target`) that handles most queries directly though it can also delegate to attributes like `#xevm.dpas_types` (in this case constructed by `xevm.target`) if so desired. This means there should be a `XeTargetAttr` implementing the `DLTIQueryInterface`'s single `query(..)` method. We can discuss what's the best mechanism for implementing this method.

Not every query above is (easily) implemented with current upstream DLTI. I am working on a PR (oriented towards DLTI attributes that respond to queries by querying for target info and applying cost models) that will enable all the above that I will push as a draft this week.

mshahneo commented:
@tkarna , yes, we would like to.
Current plan was to utilize DLTI only to query/expose the target triple. 

As @rolfmorel pointed out, the current DLTI is too verbose and simple, @adam-smnk & I spoke about using DLTI directly, the issue was mapping all the hardware info in the DLTI attribute was making the IR way too verbose and big (e.g., think the .yaml file in the atttibute list :().

Thanks @rolfmorel for the detailed example. I remember you said that you envision that a query can essentially take a C++ function as key. That's why in the current iteration, I tried to make the utilities a first class citizen, this way, any ops that implement these interfaces can be queried via DLTI:

`%dpas_supported_M_sizes = transform.dlti.query ["xe","inst","dpas","get_suppported_M", "bf16"] at %mod  // #xevm.target answers the possible M sizes for type bf16...`

This would essentially call that specific uArch's implementation of getSupportedM() function.

I'd like to discuss more with you all on this.

adam-smnk commented on mlir/include/mlir/Dialect/XeGPU/Utils/uArch.h:
```diff
@@ -0,0 +1,314 @@
+//   dim: [2, 2]
+// This represents a 2x2 tile
+struct Tile {
+  uint no_of_dims;
+  std::vector<uint> dims;
```
What exactly is the meaning of `no_of_dims`.
In this case, if `dims.size()` is different than `no_of_dims` then what happens?

mshahneo replied:
You are right, we could just remove no_of_dims and get it from dims.size().
Maybe remove the class itself. Let me see.

adam-smnk commented on mlir/include/mlir/Dialect/XeGPU/Utils/uArch.h:
```diff
@@ -0,0 +1,314 @@
+};
+
+// An enum class to represent the type of memory
+enum class MemoryType { Shared, Local, Global, Constant, Texture, Other };
```
Just thinking out loud here

Now that `XeVM` is coming, and it captures a lot hardware semantics, there could be opportunity for reuse. Maybe these basic enums/properties could lean on XeVM definitions.

mshahneo replied:
For our current use case, you are right we could use XeVM definitions.
However, I intentionally wanted to keep it separate from XeVM primarily because, if in the future there is interest, it could potentially be used for other vendors.

adam-smnk replied:
Keeping it generic also has its advantages. No strong preferences here so let's keep it as is 👍

adam-smnk reviewed (commented):
I like the general direction 👍
The overall concept to have common descriptors (the `uArch.h`) which can have platform specific implementations (PVC, battlemage variants etc.) is great.

I haven't paid too close attention to exact implementation yet. However, it is a bit hard to really judge these helper APIs and their exact abstraction right now. I think it might be best to just start with a small prototype as soon as possible and just iterate on the go.

@chencha3 I think the on-going XeGPU verifier relaxation could great opportunity to run more hands-on experiments with this infrastructure. Perhaps a small uArch-aware validation pass for XeGPU or XeVM ops?

@tkarna If you could share what key hardware properties you already use for tuning or which further info could help you in that, it'd be super valuable feedback to help steer this design.

tkarna commented:
> @tkarna If you could share what key hardware properties you already use for tuning or which further info could help you in that, it'd be super valuable feedback to help steer this design.

In my preliminary tests I've use the following parameters for 4k matmul benchmark.

Fixed parameters:

- DPAS tile sizes (n, m, k)

Tunable parameters:

- WG tile (n, m); SG tile (n, m); K tile size
- Prefetch tile sizes for A and B
- Block (load) tile sizes for A and B

Implemented constrains: tile size must not exceed the parent tile size and divide parent tile evenly. SG tiling must not exceed max number of threads (32). Prefetch tile size must be compatible with the SG level tread configuration (coop prefetching).

There are other hardware constraints I have not explicitly expressed yet, for example maximum/supported load tile size and VNNI packing constraints for B. Some configs do spill registers and run slower.

mshahneo commented:
> I like the general direction 👍 The overall concept to have common descriptors (the `uArch.h`) which can have platform specific implementations (PVC, battlemage variants etc.) is great.
> 
> I haven't paid too close attention to exact implementation yet. However, it is a bit hard to really judge these helper APIs and their exact abstraction right now. I think it might be best to just start with a small prototype as soon as possible and just iterate on the go.
> 
> @chencha3 I think the on-going XeGPU verifier relaxation could great opportunity to run more hands-on experiments with this infrastructure. Perhaps a small uArch-aware validation pass for XeGPU or XeVM ops?
> 
> @tkarna If you could share what key hardware properties you already use for tuning or which further info could help you in that, it'd be super valuable feedback to help steer this design.

Thank you so much, @adam-smnk for the feedback :-) .
I am working on the prototype, I should be able push the changes today.

mshahneo committed:
Add necessary infrastructures for the uArch to show the full pipeline.

mshahneo committed:
Add BMG to the uArch.

mshahneo committed:
Move uArch to a new folder.
Remove compiler warnings and errors.

mshahneo committed:
Use shared_pointer across the board.

mshahneo committed:
Add xegpu-attch-target-device pass.

mshahneo committed:
Add the usage of uArch in XeGPU verification.

mshahneo committed:
Remove comments, add test cases.

mshahneo committed:
Move uArchMap population in XeGPUDialect Initialization phase.

This way it's available always.

mshahneo commented:
Hi @adam-smnk , @rolfmorel , @tkarna , @Jianhui-Li, @silee2, @chencha3 , @charithaintc,

I added modified the PR to add a prototype. Please take a look at your conveniences.

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
Am I correct that this file is supposed to replace current https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTargetInfo.h?

mshahneo replied:
Potentially, yes.

Garra1980 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -577,6 +579,45 @@ LogicalResult DpasOp::verify() {
     return emitOpError("Expecting the acc type to be the same as result.");
 
+  // @uArch: Check if the types are supported for DPAS.
+  Operation *op = getOperation();
```
I guess this piece can be hidden inside uArch utilities so client just say smth like 
`auto res = verify(op, "dpas");`
?

mshahneo replied:
The issue with this is that it requires the utility to have all the parameters for of the `op`.
Alternatively, the op would have to implement the same interface so that the uArch utility can get the parameters.

chencha3 commented on mlir/test/Dialect/XeGPU/uarch-info.mlir:
```diff
@@ -0,0 +1,14 @@
+module @eltwise_add attributes {gpu.container_module} {
```
It seems the test checks are missing here.

mshahneo replied:
Ah, sorry Chao, this test is supposed to be removed. The other 2 tests covers the mechanism. I pushed it initially forgot to remove it.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Passes.td:
```diff
@@ -71,4 +71,21 @@ def XeGPUBlocking: Pass<"xegpu-blocking"> {
+      "xegpu::XeGPUDialect", "mlir::DLTIDialect"
+  ];
+}
+
```
I want to raise the following discussion points: 
1. NV and AMD are putting nvvm-attach-target and rocdl-attach-target under the GPU dialect. Not sure whether it is better for us to follow their way too. 
2. They are attached to `GPUModuleOp`, a finer granularity than us. It may be a requirement of GPU dialect. A potential benefit is that they may be able to support the case that a ModuleOp containing multiple GPUModuleOp, with each GPUMoudleOp map to different GPUs from different vendors.
3. They used NVVMTargetAttr and ROCDLTargetAttr to store the info. Now, we have XeVMTargetAttr, can we reuse it here too? Considering we only capture the device name here, which is also captured by XeVMTargetAttr?

chencha3 replied:
Sang Ik created this PR: https://github.com/llvm/llvm-project/pull/147372. is it similar to the pass here?

akroviakov replied:
I agree, this appears similar to the xevm's attach target pass that is yet to be merged. After the merge, the `chip` string of the xevm target should be appropriate to query uarch info:
```cpp
    auto targetDeviceArch =
        mlir::xegpu::uArch::uArchMap::instance().get(targetDeviceNameStr);
```

rolfmorel replied:
Do the XeVM targets have a sensible notion of a LLVM `triple` along side that of `chip`? How about (LLVM) `features`? If so, it would probably make sense to have a `#xevm.target` attribute, a la the proposed [above](https://github.com/mshahneo/llvm-project/pull/2#issuecomment-3018748822), which implements the `LLVMTargetAttrInterface` from this WIP upstream PR: 
https://github.com/llvm/llvm-project/pull/145899/files#diff-6c2503d165a7390c955c3c4fa4a76fd1991633b5ec597a1b1fd92731e1f3684dR572

I am in the process of making `#nvvm.target` and `#rocdl.target` attributes implement this interface (in addition to the generic `#llvm.target` attr in that PR). If sensible, the XeVM dialect should probably (try to) mirror this.

rolfmorel replied:
Given this in https://github.com/llvm/llvm-project/pull/147372/files#diff-3da7b8747032bb581f119fd49037d06706443c3c5db56d4d28cc74053a9b754dR284 in the Add xevm-attach-target transform pass PR, I take it there's _a_ sensible triple.

I will have a look at making `xevm.target` also implement my interface.

mshahneo replied:
Thanks Chao, Rolf, Artem.

You are right once XeVM target attribute gets upstreamed, we should use it. This pass is here just to show the use case.

rolfmorel replied:
👍 

On that note, I found that the `#xevm.target` attr itself already got merged: https://github.com/llvm/llvm-project/blob/b9d7513bf134febe72c05a04ff20f87191d7213a/mlir/include/mlir/Dialect/LLVMIR/XeVMOps.td#L521

We can get cracking!

Jianhui-Li replied:
Does this code reflect the agreement to use #xevm.target?

mshahneo replied:
Yes, Jianhui. I removed this pass. We are using #xevm.target to get target device name now.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,228 @@
+//===--- uArch.h ---------------------------------------*- C++ -*-===//
```
It should be IntelGpuXe2.h here.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+  Tensor,
+  Matrix,
+  Load,
+  Store,
```
Do Load and Store belong to different unit?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+struct Instruction {
+  std::string name;
+  std::string description;
+  std::string opcode;
```
I am not sure whether it is ok and necessary to expose the opcode here.

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/uArch/intel_gpu_pvc.yaml:
```diff
@@ -0,0 +1,205 @@
+  dpas:
+    instruction_name: dpas
+    instruction_description: "Dot Product Accumulate Systolic (DPAS) is a matrix multiply-add operation"
+    instruction_opcode: "0x83"
```
why opcode ?

mshahneo replied:
I added some extra fields in the structs in case they become useful in the future, they are not useful in our current use-case. I can remove them if we don't think they are necessary.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+//   no_of_dims: 2
+//   dim: [2, 2]
+// This represents a 2x2 tile
+struct Tile {
```
Tile is overloaded term. It looks to me just a shape. Does there any pre-defined data type you can reuse?

mshahneo replied:
Sorry, Jianhui. I am not using it in the current version of the implementation. I kept it if we want to use it in some fashion later.

The current implementation uses std::vector.

Garra1980 replied:
@mshahneo  Looks like we should remove older version of the interface and keep the one we want to proceed with

Jianhui-Li replied:
If not used, why not remove? 
Even used, I don't think the naming is good.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+//     - [2, 16]
+// This represents a 2x2 RangeTile where the first dimension can have values
+// from 1 to 32 and the second dimension can have values from 2 to 16
+struct RangeTile {
```
why not just use DiscreteTile, if this is not used that often.

mshahneo replied:
Kept two different ways to specify the range.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+//   dims:
+//     - [1, 2, 4, 8, 16, 32]
+//     - [2, 4, 8, 16]
+// This represents a 2x2 DiscreteTile where the first dimension can have values
```
It seems that DiscreteTile represents shapes combined from lists of sizes.  How about name them ShapeCombination or something more specific to our problem and less general?

mshahneo replied:
That makes sense. Current implementation actually uses std::vector.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,426 @@
+  std::tuple<Args...> data;
+  std::function<bool(Args...)> func;
+
+  Restriction(Args... args, std::function<bool(Args...)> f)
```
The term "restriction" is also very generic. Please consider making it specific. Validate the instance belongs to a combination?

mshahneo replied:
Restriction is supposed to be model all restriction.
And a common validate() could just apply all the restrictions. However, we can always go back to specific validations. I'll think about this.

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/uArch/IntelGpuXe2.cpp:
```diff
@@ -0,0 +1,185 @@
+  return {1, 2, 3, 4, 5, 6, 7, 8};
+}
+
+std::vector<uint32_t> DPASInstruction::getSupportedK(mlir::Type type) {
```
I am a bit confused. I thought you will read the information from dlti, not hard-coded inside the code.

adam-smnk replied:
These hardware properties are static so I'd say they should be hardcoded somewhere.
Whether you can extract them from LLVM, inside DLTI query function, or as a separate helper like this (or IMEX's target info) is matter of what's most convenient.

I don't think we should be dumping whole hardware specification into IR as DLTI attributes.

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -577,6 +579,45 @@ LogicalResult DpasOp::verify() {
+    llvm::errs() << "No parent module op.\n";
+
+  // It target device info is not attched, skip the target-specific checks
+  auto targetDeviceNameAttr = dlti::query(moduleOp, {"GPU", "name"});
```
We should discuss where to check the uArch info. The uArch info could  be checked in the lowering passes, and the op definition should check general constraints (like a.m = c.m).

mshahneo replied:
Actually the way the uArch info is set up. Once the target is attached, it can be queried from everywhere (pass and dialect).

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/uArch/intel_gpu_pvc.yaml:
```diff
@@ -0,0 +1,205 @@
+
+---
+# The target architecture name.
+arch: intel_gpu_pvc
```
how is this file being loaded?

adam-smnk replied:
That's an older artifact, we shouldn't have yaml at all.

Jianhui-Li replied:
Why not remove this file then? I get so confused.

Garra1980 replied:
Yeah, the clean up is on the way, please postpone review a bit

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/uArch/intel_gpu_pvc.yaml:
```diff
@@ -0,0 +1,205 @@
+    instruction_description: "2D Block Load instruction reads a rectangular block of memory in to GRF."
+    instruction_opcode: nan
+    instruction_functional_unit: load
+    instruction_type: simd
```
General: I don't think the yaml should write following the terminology in the internal uArch spec for PVC (like simd, systolic_depth, repeat_count, vnni ...). Instead, it should written following the SPIRV extnesion (or the OCL C function) (no mentioning of simd, using m/n/k, packing).

mshahneo replied:
Hi Jianhui,

We are not using the YAML in the current version.

mshahneo committed:
Remove YAML file.
We were not using it in the current version.
Removed it to reduce confusion.

mshahneo committed:
Cleanup based on review suggestions.

Remove Tile, RangeTile, DiscreteTile structs.
Uses std::vector instead.
Remove unncessary codes that were commented out.

mshahneo committed:
Remove the usage of xegpu-attach-target-device pass.

Instead use the xevm-attach-target pass to attach target device.
Shows the use-case in XeGPU Op verification.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,227 @@
+  DPASInstruction()
+      : Instruction("dpas",                     // name
+                    "Dot Product Accumulate",   // description
+                    "0xABCD",                   // opcode
```
Why opcode is required?

mshahneo replied:
Removed.

mshahneo committed:
Remove XeGPUAttachTargetDevicePAss.

We are using xevm-attach-attribute pass to get target name.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,227 @@
+  virtual std::vector<uint32_t> getSupportedK(mlir::Type type) override;
+  virtual std::vector<uint32_t> getSupportedN(mlir::Type type) override;
+  virtual std::vector<std::pair<unsigned, unsigned>>
+  getSupportedMatrix(mlir::Type type, MatrixType matrixType) override;
```
MatrixType => MMAOpndEnum ?  It is an enumeration to indicate which operands of DPAS.  Not related to a general matrix type.

mshahneo replied:
Done.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,227 @@
+};
+
+// struct to represent Load2D/Store2D/Prefetch instruction
+struct LoadStorePrefetch2DInstruction : public Instruction {
```
consider LoadStorePrefetch2DInstruction  => 2dBlockIOInst 
Reference:
https://registry.khronos.org/OpenCL/extensions/intel/[cl_intel_subgroup_2d_block_io](https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroup_2d_block_io.html).html

mshahneo replied:
Done.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,227 @@
+  //   std::vector<std::string> supported_types;
+  std::vector<uint32_t> supported_types_bitwidth;
+  std::map<std::string, uint32_t> alignment;
+  std::vector<std::vector<uint32_t>> supported_tile_sizes;
```
tile => block to be consistent with the popular term "2dblockload"

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+};
+
+// An enum class to represent the type of memory
+enum class MemoryType { Shared, Local, Global, Constant, Texture, Other };
```
Consider using "enum" as prefix if it is an enum.  

like MemoryType => MemSpaceEnum.

Jianhui-Li replied:
Using "Type", then reader don't know what fields exactly the type contains. Enum with a proper name helps readability.

Jianhui-Li replied:
BTW, Isn't memspace part of MLIR operation definition already?

mshahneo replied:
Removed.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+enum class MemoryType { Shared, Local, Global, Constant, Texture, Other };
+
+// An enum class to represent the memory access type
+enum class MemoryAccessType { Read, Write, ReadWrite, Other };
```
MemAccessTypeEnum?

Jianhui-Li replied:
Also, isn't this info avaialbe from theOP semantics already?  Load_nd read, Store_nd write.  Why we need it?

mshahneo replied:
Removed.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+enum class MemoryAccessType { Read, Write, ReadWrite, Other };
+
+// An enum class to represent the type of an instruction
+enum class InstructionType { SIMT, SIMD, SPMD, MIMD, Other };
```
Why we need this? I can't think of any scenario this thing can be taken lightly as perf knobs. Usually the software needs to be rewritten if the execution model changes.

Jianhui-Li replied:
Can't think this is a per instruction attribute.

mshahneo replied:
Removed.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+enum class InstructionType { SIMT, SIMD, SPMD, MIMD, Other };
+
+// An enum class to represent the scope of an instruction
+enum class InstructionScope {
```
What is the reason the instruction itself doesn't indicate the scope?

mshahneo replied:
For user to know at what scope the instruction is native to (work-item, subgroup or workgroup)

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+  Subgroup,
+  Workgroup,
+  Cluster,
+  Thread, // For CPU
```
XeGUP uArch includes enum for CPU?

mshahneo replied:
Idea was to make is generic. Removed now.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+struct uArchHierarchyComponent {
+  std::string name = ""; // optional name of the hierarchy component
+  // no. of lower hierarchy component it contains, e.g., for PVC XeCore it
+  // contains 8 threads, so no_of_component=8
```
Here in the comment that threads means 8 EU Threads? Or we 8 EUs (each run 4 threads under large register mode)?

mshahneo replied:
Yes.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+  Vector, // 1-D vector
+  Matrix,
+  Tile,
+  Other
```
what is tile? avoid other in general - it mean very little to user.

mshahneo replied:
Removed.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+// The uArch includes:
+// - the name of the uArch,
+// - the description of the uArch,
+// - the range of tiles supported by the uArch,
```
I don't think there is such a uArch level block size support. These are always per instructions. 2dload, 2dstore, and DPAS operations support different block size.

mshahneo replied:
Sorry, it was a mistake.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+  virtual ~TileOpInterface() = default;
+};
+
+enum class MatrixType { MatrixA, MatrixB, MatrixC, MatrixD };
```
MatrixType => MMAMatrixEnum

mshahneo replied:
Fixed now.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+// };
+
+// Create a TileLikeOp Interface
+struct TileOpInterface {
```
TileOpInterface => 2DBlockIOInterface

mshahneo replied:
Fixed.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+
+enum class MatrixType { MatrixA, MatrixB, MatrixC, MatrixD };
+struct MatrixOpInterface {
+  virtual bool checkSupportedMMATypes(mlir::Type AType, mlir::Type BType,
```
missing interface for check whether combination of matrix shape is supported.

mshahneo replied:
Added.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+  virtual std::vector<uint32_t> getSupportedK(mlir::Type type) = 0;
+  virtual std::vector<uint32_t> getSupportedN(mlir::Type type) = 0;
+  virtual std::vector<std::pair<unsigned, unsigned>>
+  getSupportedMatrix(mlir::Type type, MatrixType matrixType) = 0;
```
getSupportedMatrix => getSupportedShape

mshahneo replied:
Done.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h:
```diff
@@ -0,0 +1,389 @@
+  // Can provide load/store/prefetch ops supported tile sizes for a specific
+  // uarch
+  virtual std::vector<std::vector<uint32_t>>
+  getSupportedTiles(mlir::Type type) = 0;
```
getSupportedTiles => getSupportedShape

Jianhui-Li replied:
Consider each interface has four basic functions (PVC): 
getSupportedShape   - used in blocking analysis to set the right xegpu_layout.inst_data parameter 
CheckSupportedType  - used in unrolling to verify input IR 
CheckSupportedShapes  - used in unrolling to inst_data size of output IR
Validate - used in XeGPU and XeVM op verification. most heavy one - check more than type/shape. can't guarantee if certain value is only available at runtime (base ptr dynamically-allocated, dynamic shape)

Jianhui-Li replied:
consider specific interface for :  chunked load, 1dload.

mshahneo replied:
Done.

Jianhui-Li commented:
After reviewing this PR with Abadullah, my understanding is that the team developed a consensus that XeGPU uArch definition won't use DLTI, at least for now. @rolfmorel  @adam-smnk @Rengolin
I am suggesting Abadullah to clarify the uArch function interface to XeGPU passes, so that underneath implementation could switch to DLTI when it is possible. @Garra1980 
Meanwhile, I have some clarification questions about the scope of DLTI. Isn't it intended to support uArch definition?  If yes, what is the limitation and requirement to DLTI so that we could still converged to the same utility to access uArch info?

mshahneo committed:
Address changes based on Core-MLIR team suggestions.

Changes:
- Add common 5 APIs for all the interfaces.
- Make OpInterfaces more specific (e.g., 2D, 1D Block IO have seperate interfaces).
- Make the design more specific to Intel hardware.
- Remove the information we don't use, i.e., make the design simpler

mshahneo committed:
Add proper encapsulation.

mshahneo commented:
Hi @Jianhui-Li ,
Updated the PR based on our discussions.

mshahneo committed:
Fix a minor comment.

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -575,6 +579,62 @@ LogicalResult DpasOp::verify() {
+  Operation *op = getOperation();
+
+  // Use XeVM target
+  auto gpuModuleOp = op->getParentOfType<gpu::GPUModuleOp>();
```
can we wrap these code as an XeGPU utility function? say getXeGPUChipStr()

chencha3 replied:
+1, I think it is worth to have a getXeArch() to hide defining PVCuArch and BMGuArch instantiation at the beginning. It sounds not scalable.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+// For example, a restriction that checks if the number of dimensions in a
+// std::vector<std::vector<uint32_t>> is 2 can be represented as:
+// std::vector<std::vector<uint32_t>> rt =
+// {{1, 32}, {2, 16}}; Restriction<std::vector<std::vector<uint32_t>>> r1(rt,
```
nit: a format issue? Happy to see the restriction abstraction.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+  std::string name = ""; // optional name of the hierarchy component
+  // no. of lower hierarchy component it contains, e.g., for PVC XeCore it
+  // contains 8 threads, so no_of_component=8
+  uint32_t no_of_component;
```
nit: num_components? what is the major use of it?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+        num_banks(num_banks), bank_size(bank_size) {}
+
+  // Get methods
+  uint32_t getSize() const { return size; }
```
nit: maybe getSizeInBits?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+
+  const std::vector<std::string> &getModes() const { return mode; }
+
+  const std::vector<uint32_t> &getNumRegsPerThreadPerMode() const {
```
I am not sure whether PerThread PerMode is necessary in the naming num regs.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+  uint32_t size;
+  uint32_t line_size;
+  // At which component level the cache is shared
+  uArchHierarchyComponent component;
```
is component here meaning the owner, holder or cache level?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,333 @@
+//     SharedMemory shared_memory;
+// };
+
+struct uArchMap {
```
It may be worth to add some description for uArchMap, what is key here?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,223 @@
+namespace xegpu {
+namespace uArch {
+namespace Xe2Plus {
+struct XeCoreInfo {
```
can XeCoreInfo be used by other archs?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,223 @@
+  // MemoryAccessType memory_access_type;
+  //   std::vector<std::string> supported_types;
+  std::vector<uint32_t> supported_types_bitwidth;
+  std::map<std::string, uint32_t> alignment;
```
what is the string used for in alingment? the type name?

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,223 @@
+  }
+};
+
+namespace PVCuArch {
```
nit: maybe we can remove PVCuArch namespace if it is not hard required.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,223 @@
+              assert(tile.size() == 2);
+              return tile[1] * array_len *
+                         (dataType.getIntOrFloatBitWidth() / 8) <=
+                     64;
```
nit: format issue. 
Is it also available for store2d, which doesn't have array length support in, e.g., pvc
